File: dpTest/接雨水.cpp


#include "iostream"
#include "bits/stdc++.h"

using namespace std;

class Solution {
public:
    int trap(vector<int>& height) {

    	int n = height.size();

    	vector<int> left(n);
    	vector<int> right(n);

    	left[0] = height[0];
    	right[n-1] = height[n-1];

    	for(int i=1;i<n;i++) {
    		left[i] = max(left[i-1],height[i]);
    		right[n-1-i] = max(right[n-i],height[n-1-i]);
    	}

    	int ans = 0;
    	for(int i=1;i<n;i++) {
    		ans += min(left[i],right[i])-height[i];
    	}

    	return ans;
    }

};

int main() {

    vector<int> nums {0,1,0,2,1,0,1,3,2,1,2,1};
    Solution s2;
    cout<<s2.trap(nums)<<endl;
    return 0;
}

File: AAAA_TOP200/接雨水.java
package com.AAAA.www;

/**
 * @program: Pro1->C
 * @description: 基本测试文件
 * @author: MaYuan
 * @create: 2019-12-24 19:18
 * @version: 1.0
 **/

public class Rain {
    public  static int trap(int[] height) {
        int ans = 0;
        int left = 0, right = height.length - 1;
        int leftMax = 0, rightMax = 0;
        while (left < right) {
            leftMax = Math.max(leftMax, height[left]);
            rightMax = Math.max(rightMax, height[right]);
            if (height[left] < height[right]) {
                ans += leftMax - height[left];
                ++left;
            } else {
                ans += rightMax - height[right];
                --right;
            }
        }
        return ans;
    }

    public static void main(String[] args) {

        int[] nums = {0,1,0,2,1,0,1,3,2,1,2,1};
        System.out.println(trap(nums));
    }
}